Add an Explainer document to the repository.

rakuco opened the pull request:
This is a short summary of what is already in the spec. The document is
required as part of the TAG review process, which itself is part of #270.

Fixes #284.

Commit:
Add an Explainer document to the repository.

This is a short summary of what is already in the spec. The document is
required as part of the TAG review process, which itself is part of #270.

Fixes #284.

rakuco commented:
All: please take a look, writing a short description of what's already spec'ed required some effort.

I've borrowed some content from both the spec itself as well as @tomayac's [web.dev](https://web.dev/wakelock/) article about this API, I hope that's OK.

@kenchris mentioned the document might be lacking a comparison with other designs and alternatives. I didn't work on the spec back when other designs were considered (if any), so any help here is appreciated. For now, I've only mentioned NoSleep.js.

marcoscaceres commented on explainer.md:
```diff
@@ -0,0 +1,107 @@
+While this is fine most of the time, some applications need to keep the screen
+awake to complete their work. For example:
+
+* A recipe app that keeps the screen on while you bake a cake or cook dinner.
```
Just one generalized example is better. I'd also steer clear of "app", as it implies some kind of install... this API is designed to be just a generally applicable thing. 

```suggestion
* A recipe website that keeps the screen on while you are cooking.
```

rakuco replied:
Done.

marcoscaceres commented on explainer.md:
```diff
@@ -0,0 +1,107 @@
+* A recipe app that keeps the screen on while you bake a cake or cook dinner.
+* A boarding pass or ticket app that keeps the screen on until the barcode has
+  been scanned.
+* A kiosk-style app that keeps the screen on continuously.
```
I don't think this is a valid use case for the API. 

```suggestion
```

This is best handled at the OS level.

kenchris replied:
Actually, at Maersk they are using it for exactly this reason, for dashboards etc. You could say that they should control that at OS level, but sometimes it is hard to educate users to make those changes or even get the admin rights. Also you might not want people to always turn off the screen sleep, because it should only work when the app is running.

anssiko replied:
That use case from Maersk sounds reasonable to me.

@marcoscaceres can you expand why it is not valid in your opinion?

larsgk replied:
That's exactly what we use it for (kiosk-style dashboard) - but also considering it for always on monitoring screens/devices on vessels.

marcoscaceres replied:
Ok, we should clarify this a bit then. It sounds like a web apps on a kiosk specific device.

reillyeon replied:
I think there is a spectrum of kiosk-style devices. Some hardware is specifically designed for use as a kiosk and may include built-in functionality for keeping the screen on all the time or on a fixed schedule. On the other hand some "kiosks" are more ad-hoc and simply reuse commodity hardware and software. I think it is reasonable for any site that presents dashboards to have the ability to ask the browser to keep the screen on because that is frequently what dashboards are used for.

marcoscaceres replied:
Ok, happy for us to mention this. But just noting that this will likely be controversial and/or put off potential implementers.

marcoscaceres commented on explainer.md:
```diff
@@ -0,0 +1,107 @@
+awake to complete their work. For example:
+
+* A recipe app that keeps the screen on while you bake a cake or cook dinner.
+* A boarding pass or ticket app that keeps the screen on until the barcode has
+  been scanned.
```
```suggestion
 * A web page that presents a bar code, which has to be physically scanned by another person.
```

rakuco replied:
Done.

marcoscaceres commented on explainer.md:
```diff
@@ -0,0 +1,107 @@
+* A boarding pass or ticket app that keeps the screen on until the barcode has
+  been scanned.
+* A kiosk-style app that keeps the screen on continuously.
+* A web-based presentation app that keeps the screen on during a presentation.
```
This one is ok to keep, but calls into question wether we need some kind of integration with the Fullscreen API. 

For example:

```JS
body.requestFullscreen({wakelock: "screen"})
```

tomayac replied:
Or also the [Presentation API](https://w3c.github.io/presentation-api/)?

reillyeon replied:
As I understand it the current user agent behavior of holding a wake lock while a <video> tag is playing is unspecified. From a developer ergonomics perspective seems reasonable to add well-specified wake lock integrations to APIs where a wake lock may be frequently requested but I'll push back on the implication that a generic screen wake lock API is unnecessary because of those integrations.

marcoscaceres replied:
I think both API options might be nice. This API could handle the "cooking site" use case, while Presentation and Fullscreen API could handle the "presentation" use case. I filed https://github.com/whatwg/fullscreen/issues/185

marcoscaceres commented on explainer.md:
```diff
@@ -0,0 +1,107 @@
+they want to acquire a wake lock so that browsers can provide better automated
+and manual controls to prevent intentional and incidental issues.
+
+## Non-goals
+A previous iteration of this spec included both Screen and System wake locks.
+The latter has been moved to a separate spec, and is out of scope for this
+explainer.
```
I don't think we need to mention this at all, tbh. 

```suggestion
```

reillyeon replied:
Why isn't this historical context useful? I think it's important for current specifications to reference previously tried ideas as "alternatives considered" or "non-goals" to avoid repeating old discussions endlessly.

marcoscaceres replied:
I was thinking we should have this in the spec, tbh.

anssiko replied:
I think non-goals section is essential for this explainer given the historical context, see also https://github.com/w3ctag/w3ctag.github.io/blob/master/explainers.md#non-goals

That said, including similar text in the spec as well might also help readers who lack the historical context.

marcoscaceres commented on explainer.md:
```diff
@@ -0,0 +1,107 @@
+explainer.
+
+## API
+The API surface is not big: users can request a screen wake lock; if the
```
"The API surface is not big:" is a bit subjective... 

```suggestion
Users can request a screen wake lock, and if the
```

rakuco replied:
Done.

marcoscaceres commented on explainer.md:
````diff
@@ -0,0 +1,107 @@
+
+```js
+const lock = await navigator.wakeLock.request("screen");
+
+lock.addEventListener("release", () => {
+  console.log(`Has the lock been released? ${lock.released}`);
+});
+await lock.release();
````
The `lock.released` check might be confusing, because the `lock.released` will always be `true` if the event was fired. The intended use case is if the developer is holding a reference to the lock and wants to release it.  

```suggestion

lock.addEventListener("release", doSomething);

// Later, you can check it and release it...
if (lock.released === false) {
    await lock.release();
}  

```

rakuco replied:
Done.

marcoscaceres commented on explainer.md:
````diff
@@ -0,0 +1,107 @@
+await lock.release();
+```
+
+### WakeLock
+The `WakeLock` object is part of `Navigator`. It has a single method,
+`request(type)`. The `type` argument is of the `WakeLockType` enum type, which
+right now has a single value, _"screen"_.
+
+`WakeLock.request()` performs permission checks, verifies that the document
+issuing the request is visible and, if all checks succeed, asynchronously
+requests a lock from the underlying platform to prevent the screen from turning
+off.
+
+A successful `WakeLock.request()` invication returns a promise that will
+ultimately resolve with a `WakeLockSentinel` object.
+
````
The statement "The `WakeLock` object **is part of** `Navigator`" doesn't make much sense, tbh. Nothing is actually "part of" Navigator.  It doesn't really make sense to talk about the `WakeLock` and `Navigator` interfaces at all, but only how they are access. 

Additionally, this goes into details that are only really relevant for browser engineers (e.g., enum values)

```suggestion
### Requesting permission to use WakeLocks
To use the API, one must request access via `navigator.wakeLock.request("screen")`. This returns a promise which, if allowed by the user, resolves with a `WakeLockSentinel`. 

```JS
let sentinel;

try {
  sentinel = await navigator.wakeLock.request("screen")
} catch (err) {
  // we got denied access or something went wrong...
}
```

```

rakuco replied:
Done.

marcoscaceres commented on explainer.md:
```diff
@@ -0,0 +1,107 @@
+
+`WakeLockSentinel` has two read-only attributes for introspection: `released`
+indicates whether the lock has already been released, `type` returns the lock's
+type (it is always _"screen"_ at the moment).
```
```suggestion
type (which is always _"screen"_).
```

marcoscaceres commented on explainer.md:
```diff
@@ -0,0 +1,107 @@
+indicates whether the lock has already been released, `type` returns the lock's
+type (it is always _"screen"_ at the moment).
+
+The `WakeLockSentinel.release()` method is used to release the platform wake
```
Avoid `WakeLockSentinel.release()`, otherwise it looks like you are talking about a static method. 

```suggestion
The `WakeLockSentinel`'s `.release()` method is used to release the platform wake
```

More accurate is `WakeLockSentinel.prototype.release()`, which makes it clear is on the instance.

rakuco replied:
Done.

marcoscaceres commented on explainer.md:
```diff
@@ -0,0 +1,107 @@
+type (it is always _"screen"_ at the moment).
+
+The `WakeLockSentinel.release()` method is used to release the platform wake
+lock. It asynchronously emits a _"release"_ event that users can listen for in
```
```suggestion
lock. It asynchronously dispatches a _"release"_ event that one can listen for in
```

rakuco replied:
Done.

marcoscaceres commented on explainer.md:
```diff
@@ -0,0 +1,107 @@
+order to know that the lock is no longer being held.
+
+The lock might also be released automatically by the platform without
+`WakeLockSentinel.release()` being called (for example, when the page is no
```
```suggestion
`.release()` being called (for example, when the page is no
```

rakuco replied:
Done.

marcoscaceres commented on explainer.md:
```diff
@@ -0,0 +1,107 @@
+longer visible). The _"release_" event is still emitted the same way.
+
+## Security considerations
+* To avoid possible user fingerprinting issues, `WakeLock.request()` does not
```
```suggestion
* To avoid possible user fingerprinting issues, `WakeLock.prototype.request()` does not
```

rakuco replied:
Done.

marcoscaceres commented on explainer.md:
```diff
@@ -0,0 +1,107 @@
+* To avoid possible user fingerprinting issues, `WakeLock.request()` does not
+  indicate to API users whether the actual call to obtain a lock from the
+  operating system has succeeded or not. In other words, successful
+  `WakeLock.request()` calls must be treated by users as **advisory-only**.
```
"users" is confusing, either "dev"
```suggestion
  `WakeLock.prototype.request()` might be granted, but are not guaranteed to keep the screen awake.
   Ultimately, it's up to the browser and/or operating system if a lock is honored or not. 
```

rakuco replied:
Done.

marcoscaceres commented on explainer.md:
```diff
@@ -0,0 +1,107 @@
+  will be denied with a `NotAllowedError`. If the page is hidden after a lock
+  has been acquired, it will be released automatically.
+
+* Application of a wake lock causes various device components such as display
```
```suggestion
* Wake locks can cause various device components such as display
```

marcoscaceres commented on explainer.md:
```diff
@@ -0,0 +1,107 @@
+  has been acquired, it will be released automatically.
+
+* Application of a wake lock causes various device components such as display
+  or CPU to operate at higher power levels than they otherwise would. This can
+  lead to undesirable and potentially dangerous effects such as excessive
+  heating and faster than normal battery charge depletion. The latter is
+  particularly relevant to mobile devices which may not have a stationary power
```
I think say "and potentially dangerous" is going a bit overboard, tbh. Browsers are just tapping into publicly available OS APIs to provide this functionality (not arbitrary hacks that could cause a screen or device to catch on fire) - and those are hardly dangerous.  

```suggestion
  or CPU to operate at higher power levels than they otherwise would. This can
  lead to undesirable effects such as faster than normal battery charge depletion. 
  This is particularly relevant to mobile devices, which may not have a stationary power
```

rakuco replied:
Heh this was copied verbatim from the Security and privacy considerations section in the spec. I've sent #287 to reword it there too.

anssiko reviewed (approved):
Thanks @rakuco, great work with this first stab.

Commit:
Incorporate feedback from the previous iteration.

* Why
  - Stop mentioning "apps", reword the kiosk entry to make the use case more
    explicit.
* Non-goals
  - Briefly describe what the system wake lock API did.
* API
  - Do not say the API surface is not big.
  - Rework the example code.
  - Drop "Wake Lock" subsection, just mention how a lock is requested.
  - Stop referring to WakeLockSentinel.release() as if it were a static
    method.
* Security considerations
  - Also do not refer to WakeLockSentinel.release() as a static method.
  - Be more clear about how acquired locks are advisory-only.
  - Reword the item about battery consumption and battery depletion.

rakuco commented:
I've pushed a second commit incorporating some of Marcos' feedback. I've also reworded the kiosk item in "Why?", and added more context to what the System Wake Lock API did.

* Why
  - Stop mentioning "apps", reword the kiosk entry to make the use case more
    explicit.
* Non-goals
  - Briefly describe what the system wake lock API did.
* API
  - Do not say the API surface is not big.
  - Rework the example code.
  - Drop "Wake Lock" subsection, just mention how a lock is requested.
  - Stop referring to WakeLockSentinel.release() as if it were a static
    method.
* Security considerations
  - Also do not refer to WakeLockSentinel.release() as a static method.
  - Be more clear about how acquired locks are advisory-only.
  - Reword the item about battery consumption and battery depletion.

marcoscaceres commented:
Thanks @rakuco! looks great.

marcoscaceres commented:
@rakuco, if you rebase, it should fix the Travis error.

marcoscaceres commented on explainer.md:
```diff
@@ -0,0 +1,111 @@
+
+// Later, check and release the lock if necessary.
+if (lock.released === false) {
+    await lock.release();
```
```suggestion
  await lock.release();
```

rakuco replied:
Fixed, thanks.

Commit:
Fix indentation in code sample

rakuco commented:
Thanks for the review, everyone. It looks like this is finally ready for merging!
